src/Lykke.Service.IroncladDecorator/Controllers/CallbackController.cs

```diff
@@ -127,41 +127,45 @@ private async Task ProcessMobileCallback()
             await SignInToLykkeAsync(tokens.IdentityToken, userSession, tokens);
         }
 
-        private async Task<TokenData> GetTokens(string authCode)
+        private Task<TokenResponse> GetTokenResponse(string authCode)
         {
-            var httpClient = _httpClientFactory.CreateClient();
+            return _ironcladFacade.RedeemAuthorizationCodeAsync(
+                authCode,
+                Url.AbsoluteAction("SigninCallback", "Callback")
+            );
+        }
 
-            var discoveryResponse = await _discoveryCache.GetAsync();
+        private async Task<TokenData> ValidateTokenResponse(
+            AuthorizationRequestContext authorizationRequestContext,
+            TokenResponse tokenResponse)
+        {
+            var tokenResponseContext = new TokenResponseContext(tokenResponse);
 
-            if (discoveryResponse.IsError)
-            {
-                _discoveryCache.Refresh();
-                throw new Exception(discoveryResponse.Error);
-            }
+            var discoveryResponse = await _ironcladFacade.GetDiscoveryResponseAsync();
 
-            var tokenResponse = await httpClient.RequestAuthorizationCodeTokenAsync(new AuthorizationCodeTokenRequest
-            {
-                Address = discoveryResponse.TokenEndpoint,
-                Code = authCode,
-                ClientId =  _ironcladSettings.AuthClient.ClientId,
-                ClientSecret = _ironcladSettings.AuthClient.ClientSecret,
-                RedirectUri = Url.AbsoluteAction("SigninCallback", "Callback")
-            });
+            var keys = discoveryResponse.GetSecurityKeys();
+
+            var validationParameters = _tokenValidationParametersFactory.CreateTokenValidationParameters(keys);
 
-            if (tokenResponse.IsError)
-                throw new Exception(tokenResponse.Error);
+            tokenResponseContext.Validate(authorizationRequestContext, validationParameters);
 
             return new TokenData(tokenResponse);
         }
 
+        private Task SaveLykkeSession(string oldLykkeToken, TokenData tokens)
```

Comment:
Factory method, can be placed inside of LykkeSessionManager

Reply:
Fixed